src/Core/Checkout/Payment/PaymentService.php

```diff
@@ -141,6 +142,22 @@ public function finalizeTransaction(string $paymentToken, Request $request, Sale
 
         $transaction = $this->getPaymentTransactionStruct($transactionId, $context);
 
+        if ($token->isInvalidated()) {
+            // Token was already handled
+            // Check current state of the transaction to determine if we need to throw an exception
+            $stateName = $transaction->getOrderTransaction()->getStateMachineState()->getTechnicalName();
```

Comment:
`getStateMachineState` could return `null`. I guess that is what PHPStan is complaining about. Should be fixable like this:

```suggestion
            $stateName = $transaction->getOrderTransaction()->getStateMachineState()?->getTechnicalName();
```

`$stateName` would contain `null` if the StateMachineState is not present, which should then lead to throwing an exception, as we are comparing this to strings in the following lines

Reply:
that's why I added the @var tag, because it IS always present since the association is loaded in getPaymentTransactionStruct. Is adding a ? better in that case, because that's actually not true (that it can be null)

Reply:
adding the null-safe-operator is better, as it does not hurt in this case. 

the var annotation on the other side is bad. it is telling PHPStan, that this variable has this specific type. if the implementation would change, it could be possible that the annotation no longer matches the actual value. which would lead to errors, as PHPStan could not recognise this on its own. so var annotations should really be the very last resort to specify the type of a variable :slightly_smiling_face: 
You can read more about that here: https://github.com/shopware/shopware/blob/trunk/coding-guidelines/core/writing-code-for-static-analysis.md

Reply:
oh course, best would be to teach PHPStan how to read the Shopware code :sweat_smile: 
but writing a PHPStan extension that checks if an association is added to the criteria, and then knowing that the getter does not return null anymore is more than tricky :grin: